src/api-binder.ts

```diff
@@ -0,0 +1,888 @@
+		const lastReportedDependent = this.lastReportedState.dependent;
+		if (lastReportedLocal == null || lastReportedDependent == null) {
+			throw new InternalInconsistencyError(
+				`No local or dependent component of lastReportedLocal in ApiBinder.getStateDiff: ${JSON.stringify(
+					this.lastReportedState,
+				)}`,
+			);
+		}
+
+		const diff = {
+			local: _(this.stateForReport.local)
+				.omitBy((val, key: keyof DeviceApplicationState['local']) =>
+					_.isEqual(lastReportedLocal[key], val),
+				)
+				.omit(INTERNAL_STATE_KEYS)
+				.value(),
+			dependent: _(this.stateForReport.dependent)
+				.omitBy((val, key: keyof DeviceApplicationState['dependent']) =>
+					_.isEqual(lastReportedDependent[key], val),
+				)
+				.omit(INTERNAL_STATE_KEYS)
+				.value(),
+		};
+
+		return _.pickBy(diff, _.negate(_.isEmpty));
+	}
+
+	private async sendReportPatch(
+		stateDiff: DeviceApplicationState,
+		conf: { apiEndpoint: string; uuid: string },
+	) {
+		if (this.cachedBalenaApi == null) {
+			throw new InternalInconsistencyError(
+				'Attempt to send report patch without an API client',
+			);
+		}
+
+		const endpoint = url.resolve(
+			conf.apiEndpoint,
+			`/device/v2/${conf.uuid}/state`,
+		);
+
+		const requestParams = _.extend(
+			{
+				method: 'PATCH',
+				url: endpoint,
+				body: stateDiff,
+			},
+			this.cachedBalenaApi.passthrough,
+		);
+
+		await this.cachedBalenaApi._request(requestParams);
+	}
+
+	private async report() {
+		const conf = await this.config.getMany([
+			'deviceId',
+			'apiTimeout',
+			'apiEndpoint',
+			'uuid',
+			'localMode',
+		]);
+
+		if (checkTruthy(conf.localMode || false)) {
+			return;
+		}
+
+		const stateDiff = this.getStateDiff();
+		if (_.size(stateDiff) === 0) {
+			return 0;
+		}
+
+		await Bluebird.resolve(
+			this.sendReportPatch(stateDiff, conf as {
+				uuid: string;
+				apiEndpoint: string;
+			}),
+		).timeout(conf.apiTimeout as number);
+
+		this.stateReportErrors = 0;
+		_.assign(this.lastReportedState.local, stateDiff.local);
+		_.assign(this.lastReportedState.dependent, stateDiff.dependent);
+	}
+
+	private async reportCurrentState(): Promise<null> {
+		try {
+			this.reportPending = true;
+			const currentDeviceState = await this.deviceState.getStatus();
+
+			_.assign(this.stateForReport.local, currentDeviceState.local);
+			_.assign(this.stateForReport.dependent, currentDeviceState.dependent);
+
+			const stateDiff = this.getStateDiff();
+			if (_.size(stateDiff) === 0) {
+				this.reportPending = false;
+				return null;
+			}
+
+			await Bluebird.resolve(this.report()).delay(REPORT_SUCCESS_DELAY);
+			this.reportCurrentState();
+		} catch (e) {
+			this.stateReportErrors += 1;
+			this.eventTracker.track('Device state report failure', { error: e });
+			const delay = Math.min(
+				2 ** this.stateReportErrors * 500,
+				MAX_REPORT_RETRY_DELAY,
+			);
+
+			return Bluebird.delay(delay).then(this.reportCurrentState);
+		}
+
+		return null;
```

Comment:
Previously this function returned `null` and not the promise, meaning nothing would wait on it, now it returns a `Promise<null>` due to being an async function and any caller will likely wait on the promise, ie it's the equivalent of changing the previous
```cs
	_reportCurrentState: =>
		@reportPending = true
		@deviceState.getStatus()
		.then (currentDeviceState) =>
			_.assign(@stateForReport.local, currentDeviceState.local)
			_.assign(@stateForReport.dependent, currentDeviceState.dependent)
			stateDiff = @_getStateDiff()
			if _.size(stateDiff) is 0
				@reportPending = false
				return
			@_report()
			.delay(REPORT_SUCCESS_DELAY)
			.then =>
				@_reportCurrentState()
		.catch (err) =>
			@stateReportErrors += 1
			@eventTracker.track('Device state report failure', { error: err })
			delay = Math.min((2 ** @stateReportErrors) * 500, MAX_REPORT_RETRY_DELAY)
			Promise.delay(delay)
			.then =>
				@_reportCurrentState()
		return null
```
to
```cs
	_reportCurrentState: =>
		@reportPending = true
		@deviceState.getStatus()
		.then (currentDeviceState) =>
			_.assign(@stateForReport.local, currentDeviceState.local)
			_.assign(@stateForReport.dependent, currentDeviceState.dependent)
			stateDiff = @_getStateDiff()
			if _.size(stateDiff) is 0
				@reportPending = false
				return
			@_report()
			.delay(REPORT_SUCCESS_DELAY)
			.then =>
				@_reportCurrentState()
				return null
		.catch (err) =>
			@stateReportErrors += 1
			@eventTracker.track('Device state report failure', { error: err })
			delay = Math.min((2 ** @stateReportErrors) * 500, MAX_REPORT_RETRY_DELAY)
			Promise.delay(delay)
			.then =>
				@_reportCurrentState()